cluster/gce/gci/credential-provider/provider.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"k8s.io/klog/v2"
+	credentialproviderv1alpha1 "k8s.io/kubelet/pkg/apis/credentialprovider/v1alpha1"
+)
+
+const (
+	maxReadLength = 10 * 1 << 20 // 10MB
+)
+
+var containerRegistryUrls = []string{"container.cloud.google.com", "gcr.io", "*.gcr.io", "*.pkg.dev"}
+
+var metadataHeader = &http.Header{
+	"Metadata-Flavor": []string{"Google"},
+}
+
+// HTTPError wraps a non-StatusOK error code as an error.
+type HTTPError struct {
+	StatusCode int
+	URL        string
+}
+
+// Error implements error
+func (h *HTTPError) Error() string {
+	return fmt.Sprintf("http status code: %d while fetching url %s",
+		h.StatusCode, h.URL)
+}
+
+// TokenBlob is used to decode the JSON blob containing an access token
+// that is returned by GCE metadata.
+type TokenBlob struct {
+	AccessToken string `json:"access_token"`
+}
+
+type provider struct {
+	client        *http.Client
+	tokenEndpoint string
+}
+
+func (p *provider) Provide(image string) (map[string]credentialproviderv1alpha1.AuthConfig, error) {
+	cfg := map[string]credentialproviderv1alpha1.AuthConfig{}
+
+	tokenJSONBlob, err := readURL(p.tokenEndpoint, p.client, metadataHeader)
+	if err != nil {
+		return cfg, err
+	}
+
+	var parsedBlob TokenBlob
+	if err := json.Unmarshal(tokenJSONBlob, &parsedBlob); err != nil {
+		return cfg, err
+	}
+
+	authConfig := credentialproviderv1alpha1.AuthConfig{
+		Username: "_token",
+		Password: parsedBlob.AccessToken,
+	}
+
+	// Add our entry for each of the supported container registry URLs
+	for _, k := range containerRegistryUrls {
+		cfg[k] = authConfig
+	}
+	return cfg, nil
+}
+
+func readURL(url string, client *http.Client, header *http.Header) (body []byte, err error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	if header != nil {
+		req.Header = *header
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		klog.V(2).Infof("body of failing http response: %v", resp.Body)
```

Comment:
nit: this is going to be logged out of order with the error, which might be slightly confusing. Consider adding the body to the error type?

Reply:
This part of the code was mainly copied from https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/gcp/metadata.go but I think we should just remove this log because the kubelet reads straight from stdout